lib/api/CFieldDataCategorizer.cc

```diff
@@ -506,13 +515,46 @@ bool CFieldDataCategorizer::handleControlMessage(const std::string& controlMessa
     return true;
 }
 
-void CFieldDataCategorizer::acknowledgeFlush(const std::string& flushId) {
+void CFieldDataCategorizer::acknowledgeFlush(const std::string& flushId, bool lastHandler) {
     if (flushId.empty()) {
         LOG_ERROR(<< "Received flush control message with no ID");
     } else {
         LOG_TRACE(<< "Received flush control message with ID " << flushId);
     }
-    m_JsonOutputWriter.acknowledgeFlush(flushId, 0);
+    writeOutChangedCategories();
+    if (lastHandler) {
+        m_JsonOutputWriter.acknowledgeFlush(flushId, 0);
+    }
+}
+
+void CFieldDataCategorizer::writeOutChangedCategories() {
+    int numCategories = static_cast<int>(m_DataCategorizer->numCategories());
+    if (numCategories == 0) {
+        return;
+    }
+    std::string searchTerms;
+    std::string searchTermsRegex;
+    std::size_t maxLength;
+    bool wasCached(false);
+    for (int categoryId = 1; categoryId <= numCategories; categoryId++) {
+        if (m_DataCategorizer->categoryChangedAndReset(categoryId)) {
+            searchTerms.clear();
+            searchTermsRegex.clear();
+            maxLength = 0;
```

Comment:
I don't think you need to reset these 3 variables here.  They're only local variables and `m_DataCategorizer->createReverseSearch` will return `false` if it doesn't assign to them (or if it doesn't then that's the bug that should be fixed instead of working around it here).